### PR TITLE
feat(pipeline): pipeline support action custom eci

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
pipeline support action custom eci disable label

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=288345&iterationID=1115&pId=0&type=TASK)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English | pipeline support action custom eci disable label             |
| 🇨🇳 中文    |    pipeline支持action自定义是否禁止eci          |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: apistructs/extention.go [hunk 1]
```diff
@@ -103,6 +107,19 @@ func (spec *Spec) GetLocaleDesc(lang string) string {
 	return spec.Locale[lang][specDesc]
 }
 
+func (spec *Spec) IsDisableECI() (disableECI bool) {
+	if spec.Labels == nil {
+		return
+	}
+	if disableECIStr, ok := spec.Labels[ExtensionSpecDisableECILabel]; ok {
+		if disableECIStr == "true" {
```
Comment (sfwn) [inline apistructs/extention.go:115]: strconv.parsebool

File: apistructs/extention.go [hunk 2]
```diff
@@ -207,7 +224,7 @@ type ExtensionVersionGetRequest struct {
 // ExtensionVersionQueryRequest 查询extension版本
 type ExtensionVersionQueryRequest struct {
 	Name       string
-	YamlFormat bool
+	YamlFormat bool `query:"yamlFormat"`
```
Comment (sfwn) [inline apistructs/extention.go:227]: why just add query tag for this field? how about others fields?

File: modules/pipeline/conf/conf.go [hunk 3]
```diff
@@ -111,6 +111,9 @@ type Conf struct {
 
 	// k8s executor goroutine pool size
 	K8SExecutorPoolSize int `env:"K8S_EXECUTOR_POOL_SIZE" default:"50"`
+
+	// external market refresh interval
+	ExtensionVersionRefreshIntervalMinute uint64 `env:"EXTENSION_VERSION_REFRESH_INTERVAL_MINUTE" default:"20"`
```
Comment (sfwn) [inline modules/pipeline/conf/conf.go:116]: dicehub need push event like clusterinfo.

File: modules/pipeline/pkg/container_provider/container_instance_provider.go [hunk 4]
```diff
@@ -67,19 +68,29 @@ func WithLabels(labels map[string]string) Option {
 
 // WithStages if the stages contain custom-type action, then it will make a disabled container instance provider
 // todo judge the container instance type in task-level
-func WithStages(stages []*pipelineyml.Stage) Option {
+func WithStages(extSvc *extmarketsvc.ExtMarketSvc, stages []*pipelineyml.Stage) Option {
```
Comment (sfwn) [inline modules/pipeline/pkg/container_provider/container_instance_provider.go:71]: I have to say this dependency is strange.

File: modules/pipeline/services/extmarketsvc/list.go [hunk 5]
```diff
@@ -39,15 +42,87 @@ func SearchActionWithRender(placeholders map[string]string) OpOption {
...
+	defer s.Unlock()
+	if version == "" {
+		action, found = s.defaultActions[name]
+		return
+	}
+	action, found = s.actions[nameVersion]
+	return
+}
+
+func (s *ExtMarketSvc) continuousRefreshActionAsync() {
```
Comment (sfwn) [inline modules/pipeline/services/extmarketsvc/list.go:110]: continuousRefreshActions

Comment (sfwn) [reply modules/pipeline/services/extmarketsvc/list.go:110]: no `async` inside this method, `async` outside use `go` already.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/extention.go (hunk 2 above)
Comment (chengjoey) [reply apistructs/extention.go:227]: other fileds doesn't existed in query

File: apistructs/extention.go (hunk 1 above)
Comment (chengjoey) [reply apistructs/extention.go]: adjusted

File: modules/pipeline/conf/conf.go (hunk 3 above)
Comment (chengjoey) [reply modules/pipeline/conf/conf.go:116]: // todo

File: modules/pipeline/pkg/container_provider/container_instance_provider.go (hunk 4 above)
Comment (chengjoey) [reply modules/pipeline/pkg/container_provider/container_instance_provider.go]: adjusted

File: modules/pipeline/conf/conf.go (hunk 3 above)
Comment (sfwn) [reply modules/pipeline/conf/conf.go:116]: default 20 is too long. i think 1min is better before we can listen ext change events.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/services/extmarketsvc/list.go [hunk 6]
```diff
@@ -16,16 +16,23 @@ package extmarketsvc
...
 	"gopkg.in/yaml.v3"
 
 	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/pipeline/conf"
 	"github.com/erda-project/erda/pkg/parser/diceyml"
 	"github.com/erda-project/erda/pkg/parser/pipelineyml"
 )
 
+var (
+	defaultVersion = "default"
```
Comment (sfwn) [inline modules/pipeline/services/extmarketsvc/list.go:33]: unused var

File: modules/pipeline/services/extmarketsvc/list.go [hunk 7]
```diff
@@ -39,15 +46,116 @@ func SearchActionWithRender(placeholders map[string]string) OpOption {
...
+// getOrUpdateExtension get the fitted extension from the cache
+// if not exist, try to update the cache by the given extension name
+func (s *ExtMarketSvc) getOrUpdateExtension(nameVersion string) (action apistructs.ExtensionVersion, found bool) {
+	splits := strings.SplitN(nameVersion, "@", 2)
+	name := splits[0]
+	version := ""
+	if len(splits) > 1 {
+		version = splits[1]
+	}
+	s.Lock()
```
Comment (sfwn) [inline modules/pipeline/services/extmarketsvc/list.go:106]: lock after bundle invoked, otherwise maybe you will lock long time.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/services/extmarketsvc/list.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/services/extmarketsvc/list.go:33]: adjusted, is used in empty-version

**Comment by sfwn** [comment]
Add doc to explain that developer need to wait max 1min to get latest action info.

**Comment by chengjoey** [comment]
> Add doc to explain that developer need to wait max 1min to get latest action info.

[developer doc](https://yuque.antfin.com/dice/zs3zid/sznkxe)

**Review by sfwn — APPROVED** [review]

**Pull request merged**